services/users/users.go

```diff
@@ -6,28 +6,51 @@ package users
 import (
 	"context"
 	"fmt"
+	"runtime"
 
+	"github.com/mattermost/mattermost-server/v5/einterfaces"
 	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/services/cache"
 	"github.com/mattermost/mattermost-server/v5/shared/i18n"
 	"github.com/mattermost/mattermost-server/v5/shared/mlog"
 	"github.com/mattermost/mattermost-server/v5/store"
 )
 
 type UserService struct {
-	store  store.UserStore
-	config func() *model.Config
+	store        store.UserStore
+	sessionStore store.SessionStore
+	sessionCache cache.Cache
+	metrics      einterfaces.MetricsInterface
+	cluster      einterfaces.ClusterInterface
+	config       func() *model.Config
 }
 
 type UserCreateOptions struct {
 	Guest      bool
 	FromImport bool
 }
 
-func New(s store.UserStore, cfgFn func() *model.Config) *UserService {
-	return &UserService{
-		store:  s,
-		config: cfgFn,
+func New(s store.UserStore, ss store.SessionStore, c einterfaces.ClusterInterface, m einterfaces.MetricsInterface, cfgFn func() *model.Config) (*UserService, error) {
```

Comment:
This is reaching the threshold of too many params. Might be good time to have an initializer struct / functional options. 0/5.

Reply:
I liked the idea 👍 Also, what about having both? An initializer for madatory fields and functional arguments for optional parameters?

Reply:
Feels a bit odd, but no objections from me. :+1: